File: youtube-links/src/clients/PlaylistApiClient.ts
import type { PagedList } from '@/shared/abstractions/PagedList';
import type { CreatePlaylist } from '@/shared/features/playlists/commands/CreatePlaylist';
import type { ExportPlaylist } from '@/shared/features/playlists/commands/ExportPlaylist';
import type { ImportPlaylist } from '@/shared/features/playlists/commands/ImportPlaylist';
import type { ResetLinksDownloadedFlag } from '@/shared/features/playlists/commands/ResetLinksDownloadedFlag';
import type { UpdatePlaylist } from '@/shared/features/playlists/commands/UpdatePlaylist';
import type { GetAllPublicPlaylists } from '@/shared/features/playlists/queries/GetAllPublicPlaylists';
import type { GetAllUserPlaylists } from '@/shared/features/playlists/queries/GetAllUserPlaylists';
import type { PlaylistDto } from '@/shared/features/playlists/responses/PlaylistDto';
import { apiClient, type IApiClient } from './ApiClient';
import type { AxiosResponse } from 'axios';

interface IPlaylistApiClient {
  getAllUserPlaylists(query: GetAllUserPlaylists.Query): Promise<PagedList<PlaylistDto>>;
  getAllPublicPlaylists(query: GetAllPublicPlaylists.Query): Promise<PagedList<PlaylistDto>>;
  getPlaylist(id: number): Promise<PlaylistDto>;
  createPlaylist(command: CreatePlaylist.Command): Promise<void>;
  updatePlaylist(command: UpdatePlaylist.Command): Promise<void>;
  deletePlaylist(id: number): Promise<void>;
  exportPlaylist(command: ExportPlaylist.Command): Promise<AxiosResponse>;
  importPlaylistFromJson(command: ImportPlaylist.Command): Promise<void>;
  resetLinksDownloadedFlag(command: ResetLinksDownloadedFlag.Command): Promise<void>;
}

class PlaylistApiClient implements IPlaylistApiClient {
  private readonly apiClient: IApiClient;
  private readonly url: string = 'api/playlists';

  constructor() {
    this.apiClient = apiClient;
  }

  async getAllUserPlaylists(query: GetAllUserPlaylists.Query): Promise<PagedList<PlaylistDto>> {
    return await this.apiClient.post<GetAllUserPlaylists.Query, PagedList<PlaylistDto>>(
      `${this.url}/all`,
      query
    );
  }

  async getAllPublicPlaylists(query: GetAllPublicPlaylists.Query): Promise<PagedList<PlaylistDto>> {
    return await this.apiClient.post<GetAllPublicPlaylists.Query, PagedList<PlaylistDto>>(
      `${this.url}/allPublic`,
      query
    );
  }

  async getPlaylist(id: number): Promise<PlaylistDto> {
    return await this.apiClient.get<PlaylistDto>(`${this.url}/${id}`);
  }

  async createPlaylist(command: CreatePlaylist.Command): Promise<void> {
    await this.apiClient.postWithoutResponse(this.url, command);
  }

  async updatePlaylist(command: UpdatePlaylist.Command): Promise<void> {
    await this.apiClient.put(`${this.url}/${command.id}`, command);
  }

  async deletePlaylist(id: number): Promise<void> {
    await this.apiClient.delete(`${this.url}/${id}`);
  }

  async exportPlaylist(command: ExportPlaylist.Command): Promise<AxiosResponse> {
    return await this.apiClient.postReturnAxiosResponse(`${this.url}/export`, command);
  }

  async importPlaylistFromJson(command: ImportPlaylist.Command): Promise<void> {
    await this.apiClient.postWithoutResponse(`${this.url}/import`, command);
  }

  async resetLinksDownloadedFlag(command: ResetLinksDownloadedFlag.Command): Promise<void> {
    await this.apiClient.postWithoutResponse(`${this.url}/resetDownloadedFlag`, command);
  }
}

export const playlistApiClient = new PlaylistApiClient();

File: YoutubeLinks.Shared/Clients/PlaylistApiClient.cs
using YoutubeLinks.Shared.Abstractions;
using YoutubeLinks.Shared.Features.Playlists.Commands;
using YoutubeLinks.Shared.Features.Playlists.Queries;
using YoutubeLinks.Shared.Features.Playlists.Responses;

namespace YoutubeLinks.Shared.Clients;

public interface IPlaylistApiClient
{
    Task<PagedList<PlaylistDto>> GetAllUserPlaylists(GetAllUserPlaylists.Query query);
    Task<PagedList<PlaylistDto>> GetAllPublicPlaylists(GetAllPublicPlaylists.Query query);
    Task<PlaylistDto> GetPlaylist(int id);
    Task CreatePlaylist(CreatePlaylist.Command command);
    Task UpdatePlaylist(UpdatePlaylist.Command command);
    Task DeletePlaylist(int id);
    Task<HttpResponseMessage> ExportPlaylist(ExportPlaylist.Command command);
    Task ImportPlaylistFromJson(ImportPlaylist.Command command);
    Task ResetLinksDownloadedFlag(ResetLinksDownloadedFlag.Command command);
}

public class PlaylistApiClient(IApiClient apiClient) : IPlaylistApiClient
{
    private const string Url = "api/playlists";

    public async Task<PagedList<PlaylistDto>> GetAllUserPlaylists(GetAllUserPlaylists.Query query)
    {
        return await apiClient.Post<GetAllUserPlaylists.Query, PagedList<PlaylistDto>>($"{Url}/all", query);
    }

    public async Task<PagedList<PlaylistDto>> GetAllPublicPlaylists(GetAllPublicPlaylists.Query query)
    {
        return await apiClient.Post<GetAllPublicPlaylists.Query, PagedList<PlaylistDto>>($"{Url}/allPublic", query);
    }

    public async Task<PlaylistDto> GetPlaylist(int id)
    {
        return await apiClient.Get<PlaylistDto>($"{Url}/{id}");
    }

    public async Task CreatePlaylist(CreatePlaylist.Command command)
    {
        await apiClient.Post(Url, command);
    }

    public async Task UpdatePlaylist(UpdatePlaylist.Command command)
    {
        await apiClient.Put($"{Url}/{command.Id}", command);
    }

    public async Task DeletePlaylist(int id)
    {
        await apiClient.Delete($"{Url}/{id}");
    }

    public async Task<HttpResponseMessage> ExportPlaylist(ExportPlaylist.Command command)
    {
        return await apiClient.PostReturnHttpResponseMessage($"{Url}/export", command);
    }

    public async Task ImportPlaylistFromJson(ImportPlaylist.Command command)
    {
        await apiClient.Post($"{Url}/import", command);
    }

    public async Task ResetLinksDownloadedFlag(ResetLinksDownloadedFlag.Command command)
    {
        await apiClient.Post($"{Url}/resetDownloadedFlag", command);
    }
}